src/scheduler.ml

```diff
@@ -675,64 +728,38 @@ let poll ?log ?config ~once ~finally () =
       got_signal t signal;
       Exit
   in
-  let rec wait msg =
+  let wait msg =
     let old_generator = Console.get_status_line_generator () in
     set_status_line_generator
       (fun () ->
-         { message = Some (msg ^ ".\nWaiting for filesystem changes...")
+         { message = Some (msg ^ ", waiting for filesystem changes...")
          ; show_jobs = false
          });
     let res = block_waiting_for_changes () in
     set_status_line_generator old_generator;
-    match res with
-    | Exit -> Fiber.return Got_signal
-    | Continue -> main_loop ()
-  and main_loop () =
-    once ()
-    >>= fun _ ->
-    finally ();
-    wait "Success"
+    res
   in
-  let continue_on_error () =
-    if not t.cur_build_canceled then begin
-      finally ();
-      wait "Had errors"
-    end else begin
-      set_status_line_generator
-        (fun () ->
-           { message = Some "Had errors.\nKilling current build..."
-           ; show_jobs = false
-           });
-      Queue.clear t.waiting_for_available_job;
-      Process_watcher.killall Sys.sigkill;
-      let rec loop () =
-        if Event.pending_jobs () = 0 then
-          Continue
-        else
-          match Event.next () with
-          | Files_changed -> loop ()
-          | Job_completed _ -> loop ()
-          | Signal signal ->
-            got_signal t signal;
-            Exit
-      in
-      match loop () with
-      | Exit ->
-        Fiber.return Got_signal
-      | Continue ->
-        finally ();
-        main_loop ()
-    end
-  in
-  let rec loop f =
-    match run t f with
-    | Got_signal -> (Fiber.Never, None)
-    | exception Fiber.Never -> loop continue_on_error
-    | exception exn -> (exn, Some (Printexc.get_raw_backtrace ()))
+  let rec loop () =
+    let res = run_and_cleanup t once in
+    finally ();
+    match res with
+    | Ok (Ok ()) ->
+      wait "Success" |> after_wait
+    | Ok (Error Got_signal) ->
+      (Already_reported, None)
+    | Ok (Error Never) ->
+      wait "Had errors" |> after_wait
+    | Ok (Error Files_changed) ->
+      loop ()
+    | Error exn -> (exn, Some (Printexc.get_raw_backtrace ()))
```

Comment:
We should probably capture this backtrace earlier, i.e. when constructing the original `Error exn`